In [ ]:
# pip install pandas
# pip install numpy
# pip install datetime
import pandas as pd
import numpy as np
import datetime as dt

In [ ]:
import os
os.getcwd()

'/content'

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
df = pd.read_csv("/content/drive/MyDrive/DA 2025/market basket analysis.csv")
df

order_id product_code                     product_name  quantity  \
0        493410      TEST001          This is a test product.         5   
1       C493411        21539          RETRO SPOTS BUTTER DISH        -1   
2        493412      TEST001          This is a test product.         5   
3        493413        21724  PANDA AND BUNNIES STICKER SHEET         1   
4        493413        84578   ELEPHANT TOY WITH BLUE T-SHIRT         1   
...         ...          ...                              ...       ...   
461768   539991        21618   4 WILDFLOWER BOTANICAL CANDLES         1   
461769   539991        72741            GRAND CHOCOLATECANDLE         4   
461770   539992        21470    FLOWER VINE RAFFIA FOOD COVER         1   
461771   539992        22258          FELT FARM ANIMAL RABBIT         1   
461772   539992        21155            RED RETROSPOT PEG BAG         1   

                 order_date  price  customer_id  
0       2010-01-04 09:24:00   4.50      12346.0  
1       2010-01-04 09:43:00   4.25      14590.0  
2       2010-01-04 09:53:00   4.50      12346.0  
3       2010-01-04 09:54:00   0.85          NaN  
4       2010-01-04 09:54:00   3.75          NaN  
...                     ...    ...          ...  
461768  2010-12-23 16:49:00   1.25          NaN  
461769  2010-12-23 16:49:00   1.45          NaN  
461770  2010-12-23 17:41:00   3.75          NaN  
461771  2010-12-23 17:41:00   1.25          NaN  
461772  2010-12-23 17:41:00   2.10          NaN  

[461773 rows x 7 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 461773 entries, 0 to 461772
Data columns (total 7 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   order_id      461773 non-null  object 
 1   product_code  461773 non-null  object 
 2   product_name  459055 non-null  object 
 3   quantity      461773 non-null  int64  
 4   order_date    461773 non-null  object 
 5   price         461773 non-null  float64
 6   customer_id   360853 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 24.7+ MB


In [ ]:
# Salin dataframe asli
df_clean = df.copy()

# Membuat kolom date
df_clean['date'] = pd.to_datetime(df_clean['order_date']).dt.date.astype('datetime64[ns]')

# Menghapus semua baris tanpa customer_id
df_clean = df_clean[~df_clean['customer_id'].isna()]

#mengkonversi customer_id menjadi string
df_clean ['customer_id']= df_clean['customer_id'].astype(str)

# Menghapus semua baris tanpa product_name
df_clean = df_clean[~df_clean['product_name'].isna()]

# Membuat semua product_name berhuruf kecil
df_clean['product_name'] = df_clean['product_name'].str.lower()

# Menghapus baris dengan product_code atau product_name mengandung 'test'
df_clean = df_clean[(~df_clean['product_code'].str.lower().str.contains('test')) |
                    (~df_clean['product_name'].str.contains('test '))]

# menghapus baris dengan status cancelled, yaitu yang order_id-nya diawali 'C'
df_clean = df_clean[df_clean['order_id'].str[:1]!='C']

# Mengubah nilai quantity yang negatif menjadi positif
df_clean['quantity'] = df_clean['quantity'].abs()

# Menghapus baris dengan price bernilai negatif
df_clean = df_clean[df_clean['price'] > 0]

# Membuat kolom amount sebagai perkalian antara quantity dan price
df_clean['amount'] = df_clean['quantity'] * df_clean['price']

# Mengganti product_name berdasarkan product_code dengan yang paling sering muncul
most_freq_product_name = (
    df_clean.groupby(['product_code', 'product_name'], as_index=False)
    .agg(order_cnt=('order_id', 'nunique'))
    .sort_values(['product_code', 'order_cnt'], ascending=[True, False])
)
most_freq_product_name['rank'] = (
    most_freq_product_name.groupby('product_code')['order_cnt']
    .rank(method='first', ascending=False)
)
most_freq_product_name = most_freq_product_name[most_freq_product_name['rank'] == 1].drop(
    columns=['order_cnt', 'rank']
)
df_clean = df_clean.merge(
    most_freq_product_name.rename(columns={'product_name': 'most_freq_product_name'}),
    how='left',
    on='product_code'
)
df_clean['product_name'] = df_clean['most_freq_product_name']
df_clean = df_clean.drop(columns='most_freq_product_name')

# Menghapus outlier
from scipy import stats
df_clean = df_clean[(np.abs(stats.zscore(df_clean[['quantity','amount']]))<3).all(axis=1)]
df_clean = df_clean.reset_index(drop=True)
df_clean

<ipython-input-6-b8251154d255>:27: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean['quantity'] = df_clean['quantity'].abs()


order_id product_code                       product_name  quantity  \
0        493414        21844                  red retrospot mug        36   
1        493414        21533          retro spot large milk jug        12   
2        493414        37508    new england ceramic cake server         2   
3        493414       35001G               hand open shape gold         2   
4        493414        21527   red retrospot traditional teapot        12   
...         ...          ...                                ...       ...   
350087   539988        84380  set of 3 butterfly cookie cutters         1   
350088   539988       84849D              hot baths soap holder         1   
350089   539988       84849B             fairy soap soap holder         1   
350090   539988        22854        cream sweetheart egg holder         2   
350091   539988       47559B                tea time oven glove         2   

                 order_date  price customer_id       date  amount  
0       2010-01-04 10:28:00   2.55     14590.0 2010-01-04   91.80  
1       2010-01-04 10:28:00   4.25     14590.0 2010-01-04   51.00  
2       2010-01-04 10:28:00   2.55     14590.0 2010-01-04    5.10  
3       2010-01-04 10:28:00   4.25     14590.0 2010-01-04    8.50  
4       2010-01-04 10:28:00   6.95     14590.0 2010-01-04   83.40  
...                     ...    ...         ...        ...     ...  
350087  2010-12-23 16:06:00   1.25     18116.0 2010-12-23    1.25  
350088  2010-12-23 16:06:00   1.69     18116.0 2010-12-23    1.69  
350089  2010-12-23 16:06:00   1.69     18116.0 2010-12-23    1.69  
350090  2010-12-23 16:06:00   4.95     18116.0 2010-12-23    9.90  
350091  2010-12-23 16:06:00   1.25     18116.0 2010-12-23    2.50  

[350092 rows x 9 columns]

In [ ]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350092 entries, 0 to 350091
Data columns (total 9 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      350092 non-null  object        
 1   product_code  350092 non-null  object        
 2   product_name  350092 non-null  object        
 3   quantity      350092 non-null  int64         
 4   order_date    350092 non-null  object        
 5   price         350092 non-null  float64       
 6   customer_id   350092 non-null  object        
 7   date          350092 non-null  datetime64[ns]
 8   amount        350092 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(1), object(5)
memory usage: 24.0+ MB


In [ ]:
basket = pd.pivot_table(df_clean, index='order_id', columns='product_name', values='product_code', aggfunc='nunique', fill_value=0)
basket

product_name  10 colour spaceboy pen  12 ass zinc christmas decorations  \
order_id                                                                  
493414                             0                                  0   
493427                             0                                  0   
493428                             0                                  0   
493432                             0                                  0   
493433                             0                                  0   
...                              ...                                ...   
539981                             0                                  0   
539982                             0                                  0   
539985                             0                                  0   
539987                             0                                  0   
539988                             0                                  0   

product_name  12 coloured party balloons  12 daisy pegs in wood box  \
order_id                                                              
493414                                 0                          0   
493427                                 0                          0   
493428                                 0                          0   
493432                                 0                          0   
493433                                 0                          0   
...                                  ...                        ...   
539981                                 0                          0   
539982                                 0                          0   
539985                                 0                          0   
539987                                 0                          0   
539988                                 0                          0   

product_name  12 egg house painted wood  12 ivory rose peg place settings  \
order_id                                                                    
493414                                0                                 0   
493427                                0                                 0   
493428                                0                                 0   
493432                                0                                 0   
493433                                0                                 0   
...                                 ...                               ...   
539981                                0                                 0   
539982                                0                                 0   
539985                                0                                 0   
539987                                0                                 0   
539988                                0                                 0   

product_name  12 message cards with envelopes  12 mini toadstool pegs  \
order_id                                                                
493414                                      0                       0   
493427                                      0                       0   
493428                                      0                       0   
493432                                      0                       0   
493433                                      0                       0   
...                                       ...                     ...   
539981                                      0                       0   
539982                                      0                       0   
539985                                      0                       0   
539987                                      0                       0   
539988                                      0                       0   

product_name  12 pencil small tube woodland  12 pencils small tube posy  ...  \
order_id                                                            

In [ ]:
basket.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16272 entries, 493414 to 539988
Columns: 3842 entries, 10 colour spaceboy pen to zinc willie winkie  candle stick
dtypes: int64(3842)
memory usage: 477.1+ MB


# Encode DataFrame basket dengan nilai True untuk semua nilai di atas 0 dan False untuk semua nilai 0

In [ ]:
def encode(x):
    if x==0:
        return False
    if x>0:
        return True

basket_encode = basket.applymap(encode)
basket_encode

<ipython-input-10-31958d44448d>:7: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  basket_encode = basket.applymap(encode)


product_name  10 colour spaceboy pen  12 ass zinc christmas decorations  \
order_id                                                                  
493414                         False                              False   
493427                         False                              False   
493428                         False                              False   
493432                         False                              False   
493433                         False                              False   
...                              ...                                ...   
539981                         False                              False   
539982                         False                              False   
539985                         False                              False   
539987                         False                              False   
539988                         False                              False   

product_name  12 coloured party balloons  12 daisy pegs in wood box  \
order_id                                                              
493414                             False                      False   
493427                             False                      False   
493428                             False                      False   
493432                             False                      False   
493433                             False                      False   
...                                  ...                        ...   
539981                             False                      False   
539982                             False                      False   
539985                             False                      False   
539987                             False                      False   
539988                             False                      False   

product_name  12 egg house painted wood  12 ivory rose peg place settings  \
order_id                                                                    
493414                            False                             False   
493427                            False                             False   
493428                            False                             False   
493432                            False                             False   
493433                            False                             False   
...                                 ...                               ...   
539981                            False                             False   
539982                            False                             False   
539985                            False                             False   
539987                            False                             False   
539988                            False                             False   

product_name  12 message cards with envelopes  12 mini toadstool pegs  \
order_id                                                                
493414                                  False                   False   
493427                                  False                   False   
493428                                  False                   False   
493432                                  False                   False   
493433                                  False                   False   
...                                       ...                     ...   
539981                                  False                   False   
539982                                  False                   False   
539985                                  False                   False   
539987                                  False                   False   
539988                                  False                   False   

product_name  12 pencil small tube woodland  12 pencils small tube posy  ...  \
order_id                                                            

In [ ]:
basket_encode.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16272 entries, 493414 to 539988
Columns: 3842 entries, 10 colour spaceboy pen to zinc willie winkie  candle stick
dtypes: bool(3842)
memory usage: 59.7+ MB


# Ambil transaksi dengan banyaknya produk unik lebih dari 1 saja

In [ ]:
basket_filter = basket_encode[(basket_encode>0).sum(axis=1)>1]
basket_filter

product_name  10 colour spaceboy pen  12 ass zinc christmas decorations  \
order_id                                                                  
493414                         False                              False   
493427                         False                              False   
493428                         False                              False   
493432                         False                              False   
493433                         False                              False   
...                              ...                                ...   
539978                         False                              False   
539981                         False                              False   
539982                         False                              False   
539985                         False                              False   
539988                         False                              False   

product_name  12 coloured party balloons  12 daisy pegs in wood box  \
order_id                                                              
493414                             False                      False   
493427                             False                      False   
493428                             False                      False   
493432                             False                      False   
493433                             False                      False   
...                                  ...                        ...   
539978                             False                      False   
539981                             False                      False   
539982                             False                      False   
539985                             False                      False   
539988                             False                      False   

product_name  12 egg house painted wood  12 ivory rose peg place settings  \
order_id                                                                    
493414                            False                             False   
493427                            False                             False   
493428                            False                             False   
493432                            False                             False   
493433                            False                             False   
...                                 ...                               ...   
539978                            False                             False   
539981                            False                             False   
539982                            False                             False   
539985                            False                             False   
539988                            False                             False   

product_name  12 message cards with envelopes  12 mini toadstool pegs  \
order_id                                                                
493414                                  False                   False   
493427                                  False                   False   
493428                                  False                   False   
493432                                  False                   False   
493433                                  False                   False   
...                                       ...                     ...   
539978                                  False                   False   
539981                                  False                   False   
539982                                  False                   False   
539985                                  False                   False   
539988                                  False                   False   

product_name  12 pencil small tube woodland  12 pencils small tube posy  ...  \
order_id                                                            

In [ ]:
basket_filter.info()

<class 'pandas.core.frame.DataFrame'>
Index: 15004 entries, 493414 to 539988
Columns: 3842 entries, 10 colour spaceboy pen to zinc willie winkie  candle stick
dtypes: bool(3842)
memory usage: 55.1+ MB


# Mengaplikasikan apriori algorithm

# Buat list frequent itemset (kumpulan produk yang sering dibeli)

In [ ]:
pip install mlxtend

In [ ]:
# pip install mlxtend
from mlxtend.frequent_patterns import apriori

frequent_itemset = apriori(basket_filter, min_support=.01, use_colnames=True).sort_values('support', ascending=False).reset_index(drop=True)
frequent_itemset['product_cnt'] = frequent_itemset['itemsets'].apply(lambda x: len(x))
frequent_itemset

support                                           itemsets  product_cnt
0     0.177953               (white hanging heart t-light holder)            1
1     0.099440                         (regency cakestand 3 tier)            1
2     0.096841                          (jumbo bag red retrospot)            1
3     0.079912                 (pack of 72 retro spot cake cases)            1
4     0.078512                    (assorted colour bird ornament)            1
...        ...                                                ...          ...
1189  0.010064  (key fob , front  door , key fob , shed, key f...            3
1190  0.010064                            (wrap,suki and friends)            1
1191  0.010064                    (retrospot padded seat cushion)            1
1192  0.010064  (pack of 6 birdy gift tags, pack of 6 panneton...            2
1193  0.010064  (lunch bag red spotty, jumbo storage bag suki,...            3

[1194 rows x 3 columns]

In [ ]:
product_association = (
    association_rules(
        frequent_itemset,
        metric='confidence',
        min_threshold=0.7,
        num_itemsets=len(frequent_itemset)
    )
    .sort_values(by=['support', 'confidence'], ascending=[False, False])
    .reset_index(drop=True)
)

product_association


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents  \
0                  (red hanging heart t-light holder)   
1                    (sweetheart ceramic trinket box)   
2                                 (toilet metal sign)   
3                      (red retrospot sugar jam bowl)   
4                      (painted metal pears assorted)   
..                                                ...   
81  (green 3 piece mini dots cutlery set, blue 3 p...   
82  (green 3 piece mini dots cutlery set, pink 3 p...   
83     (key fob , front  door , key fob , back door )   
84  (poppy's playhouse kitchen, poppy's playhouse ...   
85  (lunch bag pink retrospot, lunch bag  black sk...   

                             consequents  antecedent support  \
0   (white hanging heart t-light holder)            0.058784   
1       (strawberry ceramic trinket box)            0.049254   
2                  (bathroom metal sign)            0.026993   
3         (red retrospot small milk jug)            0.023660   
4        (assorted colour bird ornament)            0.021927   
..                                   ...                 ...   
81  (pink 3 piece mini dots cutlery set)            0.011464   
82  (blue 3 piece mini dots cutlery set)            0.011997   
83                      (key fob , shed)            0.012463   
84          (poppy's playhouse bathroom)            0.013730   
85                (lunch bag red spotty)            0.013796   

    consequent support   support  confidence       lift  representativity  \
0             0.177953  0.042455    0.722222   4.058510               1.0   
1             0.074980  0.037457    0.760487  10.142533               1.0   
2             0.040589  0.021728    0.804938  19.831353               1.0   
3             0.037123  0.016796    0.709859  19.121592               1.0   
4             0.078512  0.016596    0.756839   9.639738               1.0   
..                 ...       ...         ...        ...               ...   
81            0.029992  0.010064    0.877907  29.271370               1.0   
82            0.030525  0.010064    0.838889  27.481853               1.0   
83            0.025060  0.010064    0.807487  32.222153               1.0   
84            0.010997  0.010064    0.733010  66.655016               1.0   
85            0.075913  0.010064    0.729469   9.609260               1.0   

    leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
0   0.031995    2.959371       0.800671  0.218525   0.662090    0.480400  
1   0.033764    3.862089       0.948103  0.431644   0.741073    0.630021  
2   0.020632    4.918499       0.975918  0.473837   0.796686    0.670121  
3   0.015917    3.318652       0.970669  0.381818   0.698673    0.581141  
4   0.014874    3.789618       0.916356  0.197933   0.736121    0.484107  
..       ...         ...            ...       ...        ...         ...  
81  0.009720    7.944827       0.977037  0.320594   0.874132    0.606731  
82  0.009698    6.017430       0.975313  0.310062   0.833816    0.584292  
83  0.009752    5.064272       0.981194  0.366505   0.802538    0.604541  
84  0.009913    3.704266       0.998709  0.686364   0.730041    0.824081  
85  0.009017    3.415821       0.908467  0.126360   0.707245    0.431021  

[86 rows x 14 columns]

1. Produk red hanging heart t-light holder dibeli dalam 5.9% dari total transaksi
2. Saat pembeli sudah membeli red hanging heart t-light holder, 72% dari mereka akan membeli White hanging heart t-light holder juga
3. Terdapat asosiasi yang kuat antara Red hanging heart t-light holder dengan White hanging heart t-light holder dengan peluang kedua produk tersebut dibeli bersamaan dalam 1 transaksi 4 kali lebih besar dari peluang kedua produk tersebut dibeli secara masing masing
4. Berdasarkan insight tersebut, produk White hanging heart t-light holder dapat didekat/ sebelah red hanging heart t-light holder atau kedua produk tersebut dapat dijual dalam bundel